x/posts/internal/keeper/keeper.go

```diff
@@ -104,84 +88,42 @@ func (k Keeper) GetPosts(ctx sdk.Context) []types.Post {
 // --- Likes
 // -------------
 
-func (k Keeper) getLikeStoreKey(id types.LikeID) []byte {
-	return []byte(types.LikeStorePrefix + id.String())
-}
-
-// GetLastLikeID returns the last like id that has been used
-func (k Keeper) GetLastLikeID(ctx sdk.Context) types.LikeID {
-	store := ctx.KVStore(k.StoreKey)
-	if !store.Has([]byte(types.LastLikeIDStoreKey)) {
-		return types.LikeID(0)
-	}
-
-	var id types.LikeID
-	k.Cdc.MustUnmarshalBinaryBare(store.Get([]byte(types.LastLikeIDStoreKey)), &id)
-	return id
-}
-
-// SetLastLikeID allows to set the last used like id
-func (k Keeper) SetLastLikeID(ctx sdk.Context, id types.LikeID) {
+// SaveLike allows to save the given like inside the store.
+// It assumes that the given like is valid.
+// If another like from the same owner and for the same post exists, returns an error.
+func (k Keeper) SaveLike(ctx sdk.Context, postID types.PostID, like types.Like) sdk.Error {
 	store := ctx.KVStore(k.StoreKey)
-	store.Set([]byte(types.LastLikeIDStoreKey), k.Cdc.MustMarshalBinaryBare(&id))
-}
+	key := []byte(types.LikesStorePrefix + postID.String())
 
-// AddLikeToPost allows to add a new like to a given post
-func (k Keeper) AddLikeToPost(ctx sdk.Context, post types.Post, like types.Like) sdk.Error {
-	// Set the correct post id inside the like
-	like.PostID = post.PostID
+	// Get the existent likes
+	var likes types.Likes
+	k.Cdc.MustUnmarshalBinaryBare(store.Get(key), &likes)
 
-	// Store the like and update the last like id
-	if err := k.SaveLike(ctx, like); err != nil {
-		return err
+	// Check for double likes
+	if likes.ContainsOwnerLike(like.Owner) {
```

Comment:
Possible to query the state before broadcasting the message?

Reply:
Unluckily, not. The `ValidateBasic` method has no access to the state.

On the other hand, we can move this to the handler. Do you think it would be better?